### PR TITLE
ignore /vendor directory

### LINKED ISSUE #136: Ignore /vendor directory
When using gofumpt ./... we should have same behaviour as gofmt and ignore /vendor directory. Now gofumpt modifies files inside of vendor.

### PR DESCRIPTION
**Pull request by Oiyoo**
closes #136

### CODE REVIEW

**Review by mvdan — CHANGES_REQUESTED** [review]
I'm afraid this is not enough; in your added test case, `gofumpt -w .` would still walk into the vendor folder. Please add that test case, too.

Fixing that will mean you'll need to check for "vendor" in two places, but that's just how the code is laid out in gofmt.

I'll even argue that `gofumpt -w vendor` and `gofumpt -w vendor/foo.go` *should* format vendored files, because gofumpt is explicitly told to do so. It should only be in `gofumpt -w .` that the vendor directory should be ignored. That is, we should only ignore the directory when walking a directory tree, i.e. when we'd implicitly format vendor, not explicitly.

I also think two test scripts are a bit overkill. One should be enough. The "golden" file is also redundant; you can just:

```
cp vendor/foo.go vendor/foo.go.golden
```

File: testdata/scripts/ignore-implicit-vendor.txt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	}
+
+	{
+
+		// lone comment
+
+	}
+}
+-- formatted.go.golden --
```
Comment (mvdan) [inline testdata/scripts/ignore-implicit-vendor.txt:46]: no need to keep this here; you can just

```
gofumpt orig.go.golden
cp stdout formatted.go.golden
```

File: gofmt.go [hunk 2]
```diff
@@ -193,6 +200,12 @@ func visitFile(path string, f os.FileInfo, err error) error {
 }
 
 func walkDir(path string) {
+	// reset for each arg
+	isVendorExplicit = false
+	if filepath.Base(path) == "vendor" {
+		isVendorExplicit = true
+	}
```
Comment (mvdan) [inline gofmt.go:207]: isVendorExplicit = filepath.Base(path) == "vendor"

File: testdata/scripts/ignore-implicit-vendor.txt [hunk 3]
```diff
@@ -0,0 +1,59 @@
+gofumpt -d formatted.go.golden
+! stdout .
+
+mkdir vendor
+
+# format explicit vendor dir
+cp orig.go.golden vendor/foo.go
+gofumpt -w vendor
+cmp vendor/foo.go formatted.go.golden
```
Comment (mvdan) [inline testdata/scripts/ignore-implicit-vendor.txt:9]: probably easiest to use `-l` and then check `stdout 'vendor/foo.go'`

File: testdata/scripts/ignore-implicit-vendor.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
+gofumpt -d formatted.go.golden
+! stdout .
```
Comment (mvdan) [inline testdata/scripts/ignore-implicit-vendor.txt:2]: this bit also goes away if you generate the formatted file yourself

File: testdata/scripts/ignore-implicit-vendor.txt (hunk 3 above)
Comment (mvdan) [reply testdata/scripts/ignore-implicit-vendor.txt:9]: another advantage of that approach is you don't modify files in place, so you avoid extra `cp` lines

File: gofmt.go [hunk 5]
```diff
@@ -56,6 +56,10 @@ var (
 	exitCode   = 0
 	rewrite    func(*ast.File) *ast.File
 	parserMode parser.Mode
+
+	// isVendorExplicit will be set to `true` if the caller explicitly called the
+	// vendor dir or a vendor file on each arg.
```
Comment (mvdan) [inline gofmt.go:61]: > isVendorExplicit is true if we are explicitly walking a vendor directory.

while at it, might as well call it `walkingVendorDir` for clarity.

**Review by mvdan — APPROVED** [review]
Thanks! Just some more nits.

File: testdata/scripts/ignore-implicit-vendor.txt (hunk 3 above)
Comment (Oiyoo) [reply testdata/scripts/ignore-implicit-vendor.txt]: thanks! PTAL

**Review by mvdan — APPROVED** [review]
Thanks!

**Pull request merged**